framework/src/meshgenerators/Poly2TriMeshGenerator.C

```diff
@@ -0,0 +1,409 @@
+
+    if (hole_boundary_ids.size() != _hole_ptrs.size())
+      paramError("hole_boundary_ids", "Need one hole_boundary_ids entry per hole, if specified.");
+
+    for (auto h : index_range(_hole_ptrs))
+      libMesh::MeshTools::Modification::change_boundary_id(*mesh, h + 1, h + 1 + end_bcid);
+
+    for (auto h : index_range(_hole_ptrs))
+    {
+      libMesh::MeshTools::Modification::change_boundary_id(
+          *mesh, h + 1 + end_bcid, hole_boundary_ids[h]);
+      mesh->get_boundary_info().sideset_name(hole_boundary_ids[h]) = hole_boundaries[h];
+      new_hole_bcid = std::max(new_hole_bcid, boundary_id_type(hole_boundary_ids[h] + 1));
+    }
+  }
+
+  if (isParamValid("output_boundary"))
+  {
+    const BoundaryName output_boundary = getParam<BoundaryName>("output_boundary");
+    const std::vector<BoundaryID> output_boundary_id =
+        MooseMeshUtils::getBoundaryIDs(*mesh, {output_boundary}, true);
+
+    libMesh::MeshTools::Modification::change_boundary_id(*mesh, end_bcid, output_boundary_id[0]);
+    mesh->get_boundary_info().sideset_name(output_boundary_id[0]) = output_boundary;
+
+    new_hole_bcid = std::max(new_hole_bcid, boundary_id_type(output_boundary_id[0] + 1));
+  }
+
+  bool doing_stitching = false;
+
+  for (auto hole_i : index_range(_hole_ptrs))
+  {
+    const MeshBase & hole_mesh = **_hole_ptrs[hole_i];
+    auto & hole_boundary_info = hole_mesh.get_boundary_info();
+    const std::set<boundary_id_type> & local_hole_bcids = hole_boundary_info.get_boundary_ids();
+
+    if (!local_hole_bcids.empty())
+      new_hole_bcid = std::max(new_hole_bcid, boundary_id_type(*local_hole_bcids.rbegin() + 1));
+    hole_mesh.comm().max(new_hole_bcid);
+
+    if (hole_i < _stitch_holes.size() && _stitch_holes[hole_i])
+      doing_stitching = true;
+  }
+
+  const boundary_id_type inner_bcid = new_hole_bcid + 1;
+
+  // libMesh mesh stitching still requires a serialized mesh, and it's
+  // cheaper to do that once than to do it once-per-hole
+  MeshSerializer serial(*mesh, doing_stitching);
+
+  for (auto hole_i : index_range(_hole_ptrs))
+  {
+    if (hole_i < _stitch_holes.size() && _stitch_holes[hole_i])
+    {
+      UnstructuredMesh & hole_mesh = dynamic_cast<UnstructuredMesh &>(*_hole_ptrs[hole_i]->get());
+      auto & hole_boundary_info = hole_mesh.get_boundary_info();
+
+      // Our algorithm here requires a serialized Mesh.  To avoid
+      // redundant serialization and deserialization (libMesh
+      // MeshedHole and stitch_meshes still also require
+      // serialization) we'll do the serialization up front.
+      MeshSerializer serial_hole(hole_mesh);
+
+      // It would have been nicer for MeshedHole to add the BCID
+      // itself, but we want MeshedHole to work with a const mesh.
+      // We'll still use MeshedHole, for its code distinguishing
+      // outer boundaries from inner boundaries on a
+      // hole-with-holes.
+
```

Comment:
```suggestion
      // hole-with-holes.
```

Minor annoyance (extra line) 😆